Implement bitArrayFindFirstSet()

fiam opened the pull request:
Returns the 1st bit set from a bitarray from a given starting
index. Can be used to quickly iterate over the set bits.

Will be used at least by the MAX7456 driver to iterate over the dirty chars which need to be sent to the MAX7456 to be updated.

fiam committed:
Add some unit tests for bitarray

Just to make sure the upcoming changes don't break anything

fiam committed:
Include <stdbool.h> in bitarray.h

Uses bool as a return type, so it should include its declaration

fiam committed:
Include comment indicating the bitarray functions expectations

Clarify that the array argument should 4-byte aligned

fiam committed:
Change BITARRAY_BIT_OP to use the MSB as bit[0]

This will allow us to search bitarrays using the asm CLZ
instruction, which counts leading zeroes. Unfortunately, there's
no instruction for counting trailing zeroes, but this code change
shouldn't have any impact on performance, since we're just
shifting bits the other way.

fiam committed:
Implement bitArrayFindFirstSet()

Returns the 1st bit set from a bitarray from a given starting
index. Can be used to quickly iterate over the set bits.

digitalentity commented on src/main/common/bitarray.h:
```diff
@@ -15,6 +15,21 @@
+// are zero.
+// Note that size must indicate the size of array in bytes.
+//
+// WARNING: This function only works for arrays where size is a multiple
```
Can we make it work if array size is NOT a multiple of 4?

fiam committed:
Add a proper type for bit arrays

- Define type bitArrayElement_t as uint32_t.
- Add BITARRAY_DECLARE() macro, which declares a bit array given
the number of desired bits, taking care of aligment and rounding
up its size to a multiple of 32.
- Change bit array functions to accept bitarrayElement_t* rather
than void*.
- Replace bit array declarations with BITARRAY_DECLARE() macro
- Add BITARRAY_FIND_FIRST_SET() macro to call bitArrayFindFirstSet()
without explicitely specifying its size.
- Update comments in bitarray.h to reflect these changes.

digitalentity commented on src/main/common/bitarray.h:
```diff
@@ -15,6 +15,27 @@
+
+typedef uint32_t bitarrayElement_t;
+
+#define BITARRAY_DECLARE(name, bits) bitarrayElement_t name[(bits + 31) / 32]
```
You missed the alignment attribute.

fiam replied:
uint32_t should already be aligned to 4 bytes
